### PR TITLE
Fix #4593 Fix false positive in "Rails/SaveBang" when 'save/update_attribute' is used with a 'case' statement

### LINKED ISSUE #4593: False positive in Rails/SaveBang
`Rails/SaveBang` cop claim on method `#save` in case conditions.

## Expected behavior

Rubocop should just ignore case when `#save` return value used in `case` switch.

## Actual behavior

`Use save! instead of save if the return value is not checked.`

## Steps to reproduce the problem

```ruby
class A
  def save
    true
  end
end
```

```ruby
class B
  def run
    case A.new.save
    when true
      # something
    when false
      # something else
  end
end
```

## RuboCop version

```
$ bundle exec rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by theRealNG**
Add support for 'Case', 'Logical AND' & 'Logical OR' for the
save_bang cop conditional check

SaveBang Cop reports false positives when using `save\update_attributes` with `case`, `logical and` and `logical or`.

-----------------

Before submitting the PR make sure the following are checked:

* [ ] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Used the same coding conventions as the rest of the project.
* [ ] Feature branch is up-to-date with `master` (if not - rebase it).
* [ ] Squashed related commits together.
* [ ] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests(`rake spec`) are passing.
* [ ] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -44,6 +44,7 @@
 * [#4589](https://github.com/bbatsov/rubocop/issues/4589): Fix false positive in `Performance/RegexpMatch` cop for `=~` is in a class method. ([@pocke][])
 * [#4578](https://github.com/bbatsov/rubocop/issues/4578): Fix false positive in `Lint/FormatParameterMismatch` for format with "asterisk" (`*`) width and precision. ([@smakagon][])
 * [#4285](https://github.com/bbatsov/rubocop/issues/4285): Make `Lint/DefEndAlignment` aware of multiple modifiers. ([@drenmi][])
+* [#4593](https://github.com/bbatsov/rubocop/issues/4593): Fix false positive in "Rails/SaveBangCop" when 'save/update_attribute' is used with a 'case' statement. ([@theRealNG][])
```
Comment (pocke) [inline CHANGELOG.md]: `Rails/SaveBangCop` should be `Rails/SaveBang`?

**Comment by theRealNG** [comment]
@pocke Not sure why the build is failing, could you help me

**Comment by pocke** [comment]
The failing is a known issue.  #4636

**Comment by theRealNG** [comment]
squashed the commits to a single commit

**Comment by theRealNG** [comment]
rebased with latest master

**Comment by theRealNG** [comment]
@pocke Do you need anything else to be done on this pull request?

**Comment by pocke** [comment]
> Do you need anything else to be done on this pull request?

No, looks good to me.

@bbatsov What do you think?

**Comment by bbatsov** [comment]
There's no such thing as a logical `and` and logical `or` in Ruby. I haven't checked the PR in details, but when I saw this in the title it was a red flag for me.

File: CHANGELOG.md [hunk 2]
```diff
@@ -46,6 +46,7 @@
 * [#4578](https://github.com/bbatsov/rubocop/issues/4578): Fix false positive in `Lint/FormatParameterMismatch` for format with "asterisk" (`*`) width and precision. ([@smakagon][])
 * [#4285](https://github.com/bbatsov/rubocop/issues/4285): Make `Lint/DefEndAlignment` aware of multiple modifiers. ([@drenmi][])
 * [#4634](https://github.com/bbatsov/rubocop/issues/4634): Handle heredoc that contains empty lines only in `Layout/IndentHeredoc` cop. ([@pocke][])
+* [#4593](https://github.com/bbatsov/rubocop/issues/4593): Fix false positive in "Rails/SaveBang" when 'save/update_attribute' is used with a 'case' statement. ([@theRealNG][])
```
Comment (bbatsov) [inline CHANGELOG.md]: The changelog entry doesn't match what the PR actually does.

The name of the cop and code references should have been in backticks, not quotes.

File: spec/rubocop/cop/rails/save_bang_spec.rb [hunk 3]
```diff
@@ -125,6 +125,46 @@
...
+
+      if pass
+        expect(cop.messages).to be_empty
+      else
+        expect(cop.messages)
+          .to eq(["`#{method}` returns a model which is always truthy."])
+      end
+    end
+
+    it "when using #{method} with logical AND" do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/save_bang_spec.rb]: with &&

And I guess this is also a problem with `and`

File: spec/rubocop/cop/rails/save_bang_spec.rb [hunk 4]
```diff
@@ -125,6 +125,46 @@
...
+
+      if pass
+        expect(cop.messages).to be_empty
+      else
+        expect(cop.messages)
+          .to eq(["`#{method}` returns a model which is always truthy."])
+      end
+    end
+
+    it "when using #{method} with logical OR" do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/save_bang_spec.rb]: Same here.

**Comment by bbatsov** [comment]
@theRealNG ping

**Comment by theRealNG** [comment]
@bbatsov Will work on this, this weekend

**Comment by theRealNG** [comment]
@bbatsov

**Comment by bbatsov** [comment]
👍

**Pull request merged**